lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,754 @@
+        parameters when we initialize the optimizer.
+
+        Returns:
+            List[Dict]: A list of parameter group dictionaries.
+        """
+        return [
+            {
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if not any(nd in n for nd in self.no_decay) and p.requires_grad
+                ],
+                "weight_decay": self.optimizer_init["init_args"]["weight_decay"],
+            },
+            {
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if any(nd in n for nd in self.no_decay) and p.requires_grad
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+
+    def configure_optimizers(self):
+        # the phase 0 parameters will have been set to require gradients during setup
+        # you can initialize the optimizer with a simple requires.grad filter as is often done,
+        # but in this case we pass a list of parameter groups to ensure weight_decay is
+        # not applied to the bias parameter (for completeness, in this case it won't make much
+        # performance difference)
+        optimizer = instantiate_registered_class(args=self._init_param_groups(), init=self.optimizer_init)
+        scheduler = {
+            "scheduler": instantiate_registered_class(args=optimizer, init=self.lr_scheduler_init),
+            **self.pl_lrs_cfg,
+        }
+        return [optimizer], [scheduler]
+
+    def configure_callbacks(self):
+        found_fts = [c for c in self.trainer.callbacks if isinstance(c, FinetuningScheduler)]  # type: ignore # noqa
+        if found_fts:
+            self.finetuningscheduler_callback = found_fts[0]
+        return super().configure_callbacks()
+
+
+# %% [markdown]
+# ### Our Training Sessions
+#
+# We'll be comparing three different finetuning training configurations. Every configuration in this example depends
+# upon a shared set of defaults, only differing in their respective finetuning schedules.
+#
+# | Experiment Tag    | Training Scenario Description                                          |
+# |:-----------------:| ---------------------------------------------------------------------- |
+# | ``fts_explicit``  | Training with a finetuning schedule explicitly provided by the user    |
+# | ``nofts_baseline``| A baseline finetuning training session (without scheduled finetuning)  |
+# | ``fts_implicit``  | Training with an implicitly generated finetuning schedule (the default)|
+#
+# Let's begin by configuring the ``fts_explicit`` scenario. We'll subsequently run the other two scenarios for
+# comparison.
+
+# %%
+# Let's create a finetuning schedule for our model and run an explicitly scheduled finetuning training scenario with it
+# Please see the ``FinetuningScheduler`` documentation for a full description of the schedule format
+
+
+ft_schedule_yaml = """
+0:
```

Comment:
quick question. As a user, can I specify the learning rate (optimizer defaults for each group), can I apply different schedulers to each param groups?

Reply:
Good point, I mention the relevant behavior in the FTS documentation but should add some clarification regarding that in the example. Right now when optimizer groups are added based upon the schedule, by default the lr specified for the initial finetuning phase is used as the initial lr for each additional group.

If you override [base_max_lr](https://github.com/speediedan/pytorch-lightning/blob/5356bb9dcce8c220fc6a2b3eab882aecdecbd325/pytorch_lightning/callbacks/finetuning_scheduler/fts.py#L93) to ``None``, then the initial lr for the new groups will be the first scheduled depth's lr scaled by 1e-1, but I've written it to be pretty straightforward to extend so that different initial lr's could be included in the schedule on a per-phase basis. 

Multiple schedulers can also be used, but to keep the complexity manageable until the first public version was stable, I was planning to only support single optimizer and lr_scheduler configurations explicitly in the first public version of FTS. [capture](https://github.com/speediedan/pytorch-lightning/blob/5356bb9dcce8c220fc6a2b3eab882aecdecbd325/pytorch_lightning/callbacks/finetuning_scheduler/fts.py#L440) 

Multiple schedulers could of course be used by passing a chained scheduler in the current version, but I'm thinking of deferring support for multiple schedulers attached to specific parameter groups and/or multiple optimizer to the next FTS version if it's okay with you.
Enabling the specification of a lr for each phase/group in the schedule should be pretty straightforward to add/test so I'll do so soon as I have a chance and update the example accordingly. Great suggestion as always!!!

Reply:
As discussed, I've added the ability to specify per-phase base learning rates into the latest [finetuning_scheduler_module](https://github.com/speediedan/pytorch-lightning/tree/finetuning_scheduler_module) and updated the tutorial/documentation accordingly.